File: Day24/minimum-score-after-removals-on-a-tree.java
// Time: O(N^2)  ::  Space: O(N) ==> For both approaches


class Solution {
    int[] tin, tout, sumXor;
    int timer = 0;

    int calc(int a, int b, int c) {
        return Math.max(a, Math.max(b, c)) - Math.min(a, Math.min(b, c));
    }

    public int minimumScore(int[] nums, int[][] edges) {
        int n = nums.length;
        List<Integer>[] adj = new ArrayList[n];
        for (int i = 0; i < n; ++i) adj[i] = new ArrayList<>();
        for (int[] edge : edges) {
            adj[edge[0]].add(edge[1]);
            adj[edge[1]].add(edge[0]);
        }

        tin = new int[n];
        tout = new int[n];
        sumXor = new int[n];

        dfs(0, -1, nums, adj);

        int res = Integer.MAX_VALUE;
        for (int u = 1; u < n; ++u) {
            for (int v = u + 1; v < n; ++v) {
                if (tin[v] > tin[u] && tin[v] < tout[u]) {
                    res = Math.min(res, calc(sumXor[0] ^ sumXor[u], sumXor[u] ^ sumXor[v], sumXor[v]));
                } else if (tin[u] > tin[v] && tin[u] < tout[v]) {
                    res = Math.min(res, calc(sumXor[0] ^ sumXor[v], sumXor[v] ^ sumXor[u], sumXor[u]));
                } else {
                    res = Math.min(res, calc(sumXor[0] ^ sumXor[u] ^ sumXor[v], sumXor[u], sumXor[v]));
                }
            }
        }

        return res;
    }

    void dfs(int node, int parent, int[] nums, List<Integer>[] adj) {
        tin[node] = timer++;
        sumXor[node] = nums[node];
        for (int child : adj[node]) {
            if (child != parent) {
                dfs(child, node, nums, adj);
                sumXor[node] ^= sumXor[child];
            }
        }
        tout[node] = timer;
    }
}




// Two-pass DFS Aproach
class Solution {
    int res = Integer.MAX_VALUE, sumXor = 0;

    int calc(int a, int b, int c) {
        return Math.max(a, Math.max(b, c)) - Math.min(a, Math.min(b, c));
    }

    int dfs2(int node, int parent, int firstXor, int anc, int[] nums, List<Integer>[] adj) {
        int secondXor = nums[node];
        for (int child : adj[node]) {
            if (child == parent) continue;
            secondXor ^= dfs2(child, node, firstXor, anc, nums, adj);
        }
        if (parent != anc) {
            res = Math.min(res, calc(firstXor, secondXor, sumXor ^ firstXor ^ secondXor));
        }
        return secondXor;
    }

    int dfs(int node, int parent, int[] nums, List<Integer>[] adj) {
        int firstXor = nums[node];
        for (int child : adj[node]) {
            if (child == parent) continue;
            firstXor ^= dfs(child, node, nums, adj);
        }
        for (int child : adj[node]) {
            if (child != parent) {
                dfs2(child, node, firstXor, node, nums, adj);
            }
        }
        return firstXor;
    }

    public int minimumScore(int[] nums, int[][] edges) {
        int n = nums.length;
        List<Integer>[] adj = new ArrayList[n];
        for (int i = 0; i < n; ++i) adj[i] = new ArrayList<>();
        for (int[] e : edges) {
            adj[e[0]].add(e[1]);
            adj[e[1]].add(e[0]);
        }

        for (int num : nums) sumXor ^= num;

        dfs(0, -1, nums, adj);
        return res;
    }
}

File: Day24/minimum-score-after-removals-on-a-tree.py
# Time: O(N^2)  ::  Space: O(N) ==> For both approaches


class Solution:
    def calc(self, part1: int, part2: int, part3: int) -> int:
        return max(part1, part2, part3) - min(part1, part2, part3)

    def minimumScore(self, nums: List[int], edges: List[List[int]]) -> int:
        n = len(nums)
        cnt = 0
        sumXor = [0] * n
        tin = [0] * n
        tout = [0] * n
        adj = [[] for _ in range(n)]

        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        def dfs(x: int, par: int):
            nonlocal cnt
            tin[x] = cnt
            cnt += 1
            sumXor[x] = nums[x]
            for y in adj[x]:
                if y == par:
                    continue
                dfs(y, x)
                sumXor[x] ^= sumXor[y]
            tout[x] = cnt

        dfs(0, -1)

        res = float("inf")
        for u in range(1, n):
            for v in range(u + 1, n):
                if tin[v] > tin[u] and tin[v] < tout[u]:
                    res = min(
                        res,
                        self.calc(
                            sumXor[0] ^ sumXor[u],
                            sumXor[u] ^ sumXor[v],
                            sumXor[v],
                        ),
                    )
                elif tin[u] > tin[v] and tin[u] < tout[v]:
                    res = min(
                        res,
                        self.calc(
                            sumXor[0] ^ sumXor[v],
                            sumXor[v] ^ sumXor[u],
                            sumXor[u],
                        ),
                    )
                else:
                    res = min(
                        res,
                        self.calc(
                            sumXor[0] ^ sumXor[u] ^ sumXor[v],
                            sumXor[u],
                            sumXor[v],
                        ),
                    )
        return res

class Solution:
    def calc(self, part1: int, part2: int, part3: int) -> int:
        return max(part1, part2, part3) - min(part1, part2, part3)

    def minimumScore(self, nums: List[int], edges: List[List[int]]) -> int:
        n = len(nums)
        adj = [[] for _ in range(n)]

        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        sumXor = 0
        for x in nums:
            sumXor ^= x

        res = float("inf")

        def dfs2(x: int, parent: int, firstXorVal: int, anc: int) -> int:
            secondXorVal = nums[x]
            for y in adj[x]:
                if y == parent:
                    continue
                secondXorVal ^= dfs2(y, x, firstXorVal, anc)
            if parent == anc:
                return secondXorVal
            nonlocal res
            res = min(res, self.calc(firstXorVal, secondXorVal, sumXor ^ firstXorVal ^ secondXorVal))
            return secondXorVal

        def dfs(x: int, parent: int) -> int:
            firstXorVal = nums[x]
            for y in adj[x]:
                if y == parent:
                    continue
                firstXorVal ^= dfs(y, x)
            for y in adj[x]:
                if y == parent:
                    dfs2(y, x, firstXorVal, x)
            return firstXorVal

        dfs(0, -1)
        return res
        